### ISSUE TITLE
Fluent Builder Operator

### ISSUE THREAD

**Issue by mlutze** [flix/flix#3664]
```scala
def |><(x: a, f: a -> Unit & ef): a & ef = f(x); x
```

So you can do
```scala
StringBuilder.new()
|>< StringBuilder.append("hello, ")
|>< StringBuilder.append("world")
|> StringBuilder.toString
```

**Comment by stephentetley** [comment]
Would `!>` work as a symbol now that we plan to deprecate it from function names? 

~~~ 

StringBuilder.new()
!> StringBuilder.append("hello, ")
!> StringBuilder.append("world")
|> StringBuilder.toString

~~~

I think Jonathan proposed that `StringBuilder.append` should return the StringBuilder a la Java so we could chained with `|>`, though to me this doesn't feel very "imperative functional" as the return type wouldn't be `Unit`.

For builders that make "configurations" I've found a monadic interface to be nice:

~~~

use CsvIO/CsvFormatBuilder.{toCsvFormat, setHeaders, setEscape, setDelimiter, setTrim};
toCsvFormat(
    setHeaders("Manufacturer" :: "Model" :: "Serial Number" :: Nil)
        *> setEscape('\\')
        *> setDelimiter(',')
        *> setTrim(true)
    )

~~~

The `CsvFormatBuilder` records errors, so `toCsvFormat` returns a `Result`.

**Comment by stephentetley** [comment]
... To highlight the difference between a "configuration" builder and `StringBuilder` (or Java's `ArrayList`) - with a configuration builder like `CsvFormatBuilder` above, you are unlikely to split the building between many functions so it isn't onerous being *in* the respective configuration builder monad. With `StringBuilder` you are quite likely to split the String building across many functions, forcing them all to be in a `StringBuilder` monad would be a heavy burden.

**Comment by magnus-madsen** [comment]
@stephentetley @mlutze @JonathanStarup 

To recap, is the proposal:

- [ ] Remove `!` from names since soon the effect system will have the same functionality.
- [ ] Add `def !>(x: a, f: a -> Unit & ef): a & ef = f(x); x`
- [ ] Add an expression `discard e` which returns unit and requires `e` to be non-pure.
- [ ] Enforce that `e1; e2` requires `e1` to return Unit (in Redundancy).
- [ ] Refactor the Std. Lib. to use `discard` where needed?

Finally, do we want to change some signatures in the language/lib?:

- [ ] We talked about changing `putchan` to return Unit.
- [ ] What else?

Essentially moving away from an OO-style chaining.

**Comment by mlutze** [comment]
I agree with all of the above.

**Comment by magnus-madsen** [comment]
We forget the past. How are we going to name:

- `Array.reverse` and `Array.reverse!`
- `Array.replace` and `Array.replace!`

**Comment by JonathanStarup** [comment]
python uses sort/sorted. the imperative sort is imperative/inplace and sorted returns a new array. for reverse this would be reverse (inplace) and reversed (functional).

I think I prefer stephens suggestion but it's something to think about.

**Comment by JonathanStarup** [comment]
Java Arrays only have inplace. that approach would be to only have inplace functions and then use copy if you don't want it inplace

**Comment by JonathanStarup** [comment]
https://www.scala-lang.org/api/2.12.15/scala/collection/mutable/ArraySeq.html#reverse:Repr


Scala seems do just support one for each operation with no preference.

**Comment by JonathanStarup** [comment]
I think this issue is partially solved by #3694 and superseded by #3771
